Add Retry-After header with backoff and jitter to active-series-limiting and quorum-unavailable responses

roth-wine opened the pull request:
* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

By default, active-series-limiting returns only a `429` when a tenant exceeds the limit, and a quorum failure returns only a `503`. Prometheus [parses the `Retry-After` header natively](https://github.com/prometheus/prometheus/blob/9ec59baffb547e24f1468a53eb82901e58feabd8/storage/remote/client.go#L322-L323). Without it, the sender falls back to its own remote-write backoff, which starts at `MinBackoff` (30ms) and doubles up to [`MaxBackoff` (5s by default)](https://github.com/prometheus/prometheus/blob/9ec59baffb547e24f1468a53eb82901e58feabd8/config/config.go#L239-L240) with no jitter. So the first retries come in fast, and with no jitter many senders that hit the limit at the same moment keep retrying in lockstep, which is what piles load onto an already constrained receiver.

This change sets the `Retry-After` header on:
- `429`: tenant has exceeded the active-series limit (honoured by senders that set `retry_on_http_429`)
- `503`: write quorum cannot be reached (replication unavailable)

Two flags control the behaviour, both with defaults so most users never touch them:
- `--receive.retry-after-backoff` (default `5s`): base backoff duration returned in the `Retry-After` header
- `--receive.retry-after-jitter` (default `0.5`): fraction of the backoff added as random jitter (±50% by default)

Setting jitter to `0` disables randomisation and returns the plain backoff value.

## Verification

Added two unit tests in `pkg/receive/handler_test.go`:

- `TestRetryAfterHeaderOnActiveSeriesLimitExceeded`: wires a `fakeAlwaysOverLimitSeriesLimiter` (always returns `isUnderLimit=false`) into the handler with a `10s` backoff and asserts the response carries HTTP `429` and a `Retry-After` header that is a valid HTTP date in the future.
- `TestRetryAfterHeaderOnQuorumUnavailable`: replaces all peer connections with a `fakeUnavailableRemoteWriteClient` (always returns `codes.Unavailable`) in a 3-node hashring with replication factor 3 so quorum can never be reached, and asserts handlers returning `503` carry a `Retry-After` header that is a valid HTTP date in the future.

Without the change, neither response carries a `Retry-After` header, so senders retry on their own jitter-less backoff and tend to retry in lockstep against a recovering endpoint.

roth-wine commented:
As far is a can see in the contributions for the thanos receive active series limiting stuff, maybe @GiedriusS or @saswatamcode can have a look at this?

saswatamcode reviewed (commented):
Generally LGTM, just a comment!

roth-wine commented:
@saswatamcode I have adjusted everything we have discussed before and rebased the branch to get this thing merged.
Failing tests seems flaky and are unrelated to my changes.

GiedriusS reviewed (commented):
Maybe it makes sense to always set this header eg if we return 503 when quorum cannot be reached?

roth-wine commented:
@GiedriusS Thanks for looking into this MR!
In general, i would highly appreciate your idea to set the header also for 503 status codes. But could we open a seperate Issue for your request to get this MR merged?

roth-wine committed:
receive: Add Retry-After header to active-series-limiting and quorum-unavailable responses

By default, active-series-limiting returns 429 without a Retry-After
header. Prometheus parses this header natively and without it performs
an instant retry, potentially increasing load further.

This change sets the Retry-After header on:
- 429 (active-series limit exceeded)
- 503 (quorum unavailable)

Two flags control the behaviour:
  --receive.active-series-limiting.retry-after-backoff (default: 5s)
  --receive.active-series-limiting.retry-after-jitter  (default: 0.5)

Jitter is applied via DurationWithJitter; when jitter is 0 the plain
backoff value is used directly.

Signed-off-by: roth-wine <git@roth-wine.de>

roth-wine committed:
receive: Add tests for Retry-After header on 429 and 503 responses

- TestRetryAfterHeaderOnActiveSeriesLimitExceeded: verifies the
  Retry-After header is set when a tenant exceeds the active-series
  limit (429 Too Many Requests).
- TestRetryAfterHeaderOnQuorumUnavailable: verifies the Retry-After
  header is set when replication quorum cannot be reached (503 Service
  Unavailable).

Signed-off-by: roth-wine <git@roth-wine.de>

GiedriusS commented on pkg/receive/handler.go:
```diff
@@ -532,6 +535,15 @@ func newWriteResponse(seriesIDs []int, err error, er endpointReplica) writeRespo
+	if h.options.RetryAfterJitter <= 0 {
+		return h.options.RetryAfterBackoff
+	}
+	return util.DurationWithJitter(h.options.RetryAfterBackoff, h.options.RetryAfterJitter)
```
I think usually exponential backoff with jitter is chosen. What are the pros and cons of each, and why you chose this one in particular?

Also, I am really not a fan of adding yet another flag. It's extra cognitive load for _all_ users. Why not just stick with 5s/0.5 as the default in the constants? How the users could know what values are the "best"?

roth-wine replied:
On the backoff choice: exponential backoff needs to know how many times a given sender has already failed in a row, and that state lives on the sender, not on us. Prometheus already does it. When its remote-write client sees a `Retry-After` it sleeps for exactly the value we send ([`queue_manager.go#L2048`](https://github.com/prometheus/prometheus/blob/9ec59baffb547e24f1468a53eb82901e58feabd8/storage/remote/queue_manager.go#L2048)) and seeds its own next backoff from it ([`queue_manager.go#L2080`](https://github.com/prometheus/prometheus/blob/9ec59baffb547e24f1468a53eb82901e58feabd8/storage/remote/queue_manager.go#L2080)). So the exponential part is already there, from the side that actually holds the retry count. If we grew the value server-side too, we'd double it up and have to track per-sender failure counts on the write path for no real gain.

What Prometheus' backoff doesn't do is jitter. It's a plain doubling capped at `MaxBackoff` ([5s by default](https://github.com/prometheus/prometheus/blob/9ec59baffb547e24f1468a53eb82901e58feabd8/config/config.go#L239-L240)), so when many senders hit the same limit at the same moment they retry in lockstep. The receiver is the only place that can spread them out, and since we recompute the jitter on every response they stay spread across retries, not just the first one.

On the flags: our setup is what made me keep them configurable. We run one receiver+distributor deployment per tenant (deliberate hard separation), and a single endpoint can have on the order of 10k senders pointed at it. When something fails (a rollout, quorum loss) they all retry, and the retry rate is basically `sender_count / backoff`, so that floor is the main thing keeping the recovering distributor from getting buried: 10k/5s is ~2000 rps, 10k/60s is ~170. Prometheus honours a `Retry-After` larger than its own `MaxBackoff` (the [sleep](https://github.com/prometheus/prometheus/blob/9ec59baffb547e24f1468a53eb82901e58feabd8/storage/remote/queue_manager.go#L2048) is set straight to our value, the [cap](https://github.com/prometheus/prometheus/blob/9ec59baffb547e24f1468a53eb82901e58feabd8/storage/remote/queue_manager.go#L2080) only applies to the fallback), so the header is the one lever we have to push senders past their client-side cap, and we don't control their remote-write config anyway.

Because each tenant is its own deployment, the global flag is effectively per-tenant for us: a big tenant gets a longer floor, a small one keeps the default. A hard-coded 5s can't cover both. Defaults stay 5s/0.5, so anyone who doesn't care never touches them.

One thing I'll fix regardless: the flags are still named under `active-series-limiting` from the original 429-only version, but they cover the 503 path now too, so I'll rename them to fit both.

roth-wine committed:
receive: drop active-series-limiting prefix from retry-after flags

The flags also set Retry-After on 503 (quorum unavailable), not only on 429 active-series-limiting, so the active-series-limiting prefix was too narrow.

Signed-off-by: roth-wine <git@roth-wine.de>

roth-wine committed:
Merge branch 'main' into pr_add_retry_after_to_active_series_limiting

sebastiangaiser commented:
@GiedriusS @saswatamcode could you please retry the two failing CI jobs. Thanks 👋🏻

GiedriusS commented on pkg/receive/handler_test.go:
```diff
@@ -2411,3 +2413,123 @@ func TestPeerGroupResetRace(t *testing.T) {
+	require.NoError(t, reqErr)
+
+	require.Equal(t, http.StatusTooManyRequests, rec.Code)
+	retryAfter := rec.Header().Get("Retry-After")
```
I could see the time checking being flaky on super limited GH actions runners. Could synctest be used instead to have a consistent time?

GiedriusS commented on pkg/receive/handler_test.go:
```diff
@@ -2411,3 +2413,123 @@ func TestPeerGroupResetRace(t *testing.T) {
+
+		retryAfter := rec.Header().Get("Retry-After")
+		require.NotEmptyf(t, retryAfter, "handler %d: Retry-After header must be set on 503", i)
+		retryTime, parseErr := http.ParseTime(retryAfter)
```
Same here - synctest please 🙏

GiedriusS commented:
Huge sorry for the delay. I read through the explanation and everything makes sense. I just think there's a way to make the tests more resilient on laggy runners through synctest. We ran into multiple issues over the years with checking the time and `synctest` should probably be always used when the tests deal with date or time.

roth-wine commented:
Hey @GiedriusS, i have moved both tests to synctest as you suggested and rebased again onto main.
Let me know if everything looks fine now :smile:

GiedriusS commented on pkg/receive/handler_test.go:
```diff
@@ -2411,3 +2413,126 @@ func TestPeerGroupResetRace(t *testing.T) {
+	require.NoError(t, err)
+	defer func() {
+		require.NoError(t, closeFunc())
+		time.AfterFunc(50*time.Millisecond, func() {
```
Is this sleep still needed after synctest changes?

roth-wine replied:
Good catch, I have removed the sleep.
